### ISSUE TITLE
Difficulty with using qKG method along with optimize_acqf_discrete and optimize_acqf_mixed functions

### ISSUE THREAD

**Issue by sambitmishra98** [pytorch/botorch#938]
I want to use optimize_acqf_discrete and optimize_acqf_mixed but am finding it difficult to use with qKG method. I was able to use qKG with optimize_acqf thanks to the documentation examples. Can you guys help me by adding a simple example for these functions in the documentation?

**Comment by eytan** [comment]
cc @danielrjiang, @dme65 , @saitcakmak

**Comment by eytan** [comment]
@sambitmishra98  you tell us a little bit more about what you tried along with a few simple reproducible examples?  Some inline code or collabs could be helpful.

**Comment by sambitmishra98** [comment]
Thanks a lot for your help on this!
I executed it in Google Colab. Below is the link: [My Code](https://colab.research.google.com/drive/10mV4bhqmcDzuOFy46Xmpm0PyhSjv6uZs?usp=sharing)

For my own practice, I basically used the following link and tried to replace `optimize_acqf` with `optimize_acqf_discrete`, hoping for the best: [Code from examples](https://botorch.org/tutorials/one_shot_kg)

For my research, I wish to implement BO to a given set of integer variables and float variables. I just want to perform BO once and obtain one candidate as output. My end goal is using `optimize_acqf_mixed` effectively, so any example using this function would help.

Thanks again!

**Comment by saitcakmak** [comment]
Hi @sambitmishra98. There's a fundamental issue with the qKG that makes it prohibitively expensive to optimize over mixed or discrete input spaces. 

In qKG, for `q` candidates and `n_f` fantasies, the solution tensor is `batch x (q + n_f) x d`-dim.  Suppose that you only have the first column of the input be discrete, and you have `k` values it can take. Since you're actually optimizing over `q + n_f` solutions jointly, you end up with `k^(q + n_f)` discrete combinations to evaluate (and optimize the continuous columns over). You can see how this blows up very fast.

Going back to your original question with `optimize_acqf_mixed/discrete`, they do not support qKG optimization properly. In `optimize_acqf_mixed`, you specify `fixed_features_list` which is a list of values a given column of the `d` dim tensor can take. Since qKG input is `(q + n_f) x d`, the same value will be applied to all q-batch dimensions, forcing both your candidates and the fantasy solutions to consider the same discrete value. It also uses sequential optimization if `q > 1`, which leads to other issues I will not discuss. 
For `optimize_acqf_discrete` with qKG, it uses sequential optimization over `q`-batches, which does not work with qKG since it requires jointly evaluating the candidates and fantasies. I don't think there's a way to get around this as it is implemented, which is reasonable since the number of combinations would blow up otherwise.

TLDR: qKG is not meant for mixed / discrete optimization (at least with the one-shot implementation). I'd recommend using some other acquisition function.

**Comment by sambitmishra98** [comment]
Okay. Maybe the error message could be made more clear on this issue? I thought it's a minor issue based on the error message I got.

I guess using the Expected Improvement (EI) method for mixed/discrete optimisation would not have the same problem? I will try to work on that. 

Thanks a lot for the explanation!

**Comment by saitcakmak** [comment]
> Maybe the error message could be made more clear on this issue? I thought it's a minor issue based on the error message I got.

Yeah, it makes sense to have a more meaningful error message there. I'll put up a PR for that.

> I guess using the Expected Improvement (EI) method for mixed/discrete optimisation would not have the same problem?

Yes, EI should be safe to use for mixed/discrete optimization. Good luck!
